Add m.ignored_user_list support

NepNep21 opened the pull request:
Testing is still absent, I can try adding it in a while. If/when this is merged I plan to make a PR to nheko, adding client support for it.

deepbluev7 commented on include/mtx/events/account_data/ignored_users.hpp:
```diff
@@ -0,0 +1,38 @@
+namespace events {
+namespace account_data {
+
+namespace data {
```
The data namespace is a bit weird, I would probably just get rid of it :)

deepbluev7 commented on lib/structs/events/account_data/ignored_users.cpp:
```diff
@@ -0,0 +1,33 @@
+void
+from_json(const nlohmann::json &obj, IgnoredUsers &content)
+{
+    const auto map = obj.at("ignored_users").get<std::unordered_map<std::string, nlohmann::json>>();
```
You should add a guard here, that the key actually exists. Some clients might remove it, if it is empty :)

deepbluev7 commented on lib/structs/events/account_data/ignored_users.cpp:
```diff
@@ -0,0 +1,33 @@
+from_json(const nlohmann::json &obj, IgnoredUsers &content)
+{
+    const auto map = obj.at("ignored_users").get<std::unordered_map<std::string, nlohmann::json>>();
+    for (const auto &[key, value] : map) {
```
I think you can just iterate over the items directly:

```
for (auto& [key, val] : obj.at("ignored_users").items())

```

deepbluev7 commented:
At a glance this looks good, but I suggest you add some tests :)

NepNep21 commented:
Tests coming soon!

NepNep21 commented:
Just added a (de)serialization test (and fixed a bug found running it, I suppose synapse was just lax enough to allow the buggy version when I manually tested), since that's what other account_data events seem to have, though I can add more if you find any necessary. Also the test feels a little janky since neither json nor nlohmann::json guarantee key ordering, so I just check whether the keys exist at all.

deepbluev7 commented on include/mtx/events/event_type.hpp:
```diff
@@ -131,6 +131,8 @@ enum class EventType
     //! m.direct
     Direct,
+    //! m.ignored_user_list
+    IgnoredUsers,
```
In general we try to name those after the event type, so this would be IgnoredUserList, but otoh it doesn't really matter here :)

NepNep21 replied:
Ahh right, sorry about that.

deepbluev7 commented on tests/events.cpp:
```diff
@@ -1615,6 +1616,38 @@ TEST(AccountData, FullyRead)
+
+    ASSERT_EQ(event.content.users.size(), 2);
+
+    // JSON does not demand a guaranteed key order
```
However our parser always sorts them alphabetically, but again, doesn't matter :D

deepbluev7 commented:
Thank you for the contribution and sorry I missed that you added the tests. Looks good to me!
